### PR TITLE
Add website data tool (Powered by Firecrawl)

### PR DESCRIPTION
**Pull request by ericciarla**
Great work on the Langchain Generative UI repository so far! We wanted to add a new feature that we think will be really useful for our users - a Website Data Tool powered by Firecrawl.

This tool will allow users to easily retrieve and display key information about a website directly within the application.

### CODE REVIEW

File: components/prebuilt/web.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
+"use client";
+
+import { Card } from "../ui/card";
+import { Skeleton } from "../ui/skeleton";
+
+export interface CurrentWebProps {
```
Comment (bracesproul) [inline components/prebuilt/web.tsx]: Nit, drop current
```suggestion
export interface WebProps {
```

File: components/prebuilt/web.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  url: string;
+  title: string;
+  description: string;
+  screenshot: string;
+}
+
+export function CurrentWebLoading() {
```
Comment (bracesproul) [inline components/prebuilt/web.tsx:13]:
```suggestion
export function WebLoading() {
```

File: components/prebuilt/web.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        <Skeleton className="h-[200px] w-[300px]" />
+      </div>
+      <div className="flex justify-between items-center mb-1">
+        <Skeleton className="h-[16px] w-full" />
+      </div>
+    </Card>
+  );
+}
+
+export function CurrentWeb(props: CurrentWebProps) {
```
Comment (bracesproul) [inline components/prebuilt/web.tsx:29]:
```suggestion
export function Web(props: WebProps) {
```

File: ai/tools/firecrawl.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import { CurrentWebLoading, CurrentWeb } from "@/components/prebuilt/web";
+import { createRunnableUI } from "@/utils/server";
+import { DynamicStructuredTool } from "@langchain/core/tools";
+import FirecrawlApp from "@mendable/firecrawl-js";
```
Comment (bracesproul) [inline ai/tools/firecrawl.tsx]: Would rather use the `FireCrawlLoader` from langchain, I can push the update.

**Review by bracesproul — COMMENTED** [review]
This all LGTM with some minor nits I'll handle pushing. Thanks for taking the time to contribute!!

File: ai/tools/firecrawl.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+export const websiteDataTool = new DynamicStructuredTool({
+  name: "get_web_data",
+  description: "A tool to fetch the current website data, given a url.",
+  schema: webSchema,
+  func: async (input, config) => {
+    const stream = await createRunnableUI(config, <CurrentWebLoading />);
+    const data = await webData(input);
+    stream.done(<CurrentWeb {...data} />);
```
Comment (bracesproul) [inline ai/tools/firecrawl.tsx:49]:
```suggestion
    const stream = await createRunnableUI(config, <WebLoading />);
    const data = await webData(input);
    stream.done(<Web {...data} />);
```

File: ai/tools/firecrawl.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import { CurrentWebLoading, CurrentWeb } from "@/components/prebuilt/web";
```
Comment (bracesproul) [inline ai/tools/firecrawl.tsx:1]:
```suggestion
import { WebLoading, Web } from "@/components/prebuilt/web";
```

**Comment by bracesproul** [comment]
Here's the PR with the naming nit & switching to the LangChain document loader: https://github.com/bracesproul/gen-ui/pull/8/files

I can't push commits to this branch, so could you merge that PR into this PR? Thanks!

File: ai/tools/firecrawl.tsx (hunk 4 above)
Comment (ericciarla) [reply ai/tools/firecrawl.tsx]: Ok awesome!

**Comment by ericciarla** [comment]
Thanks for getting this in @bracesproul! Should be all set 🔥

**Review by bracesproul — APPROVED** [review]
thank you!

**Pull request merged**